zkevm-circuits/src/sig_circuit.rs

```diff
@@ -0,0 +1,1076 @@
+                log::info!("before proceeding to the next phase");
+
+                #[cfg(not(feature = "onephase"))]
+                {
+                    // finalize the current lookup table before moving to next phase
+                    ecdsa_chip.finalize(&mut ctx);
+                    ctx.print_stats(&["ECDSA context"]);
+                    ctx.next_phase();
+                }
+
+                // ================================================
+                // step 3: compute RLC of keys and messages
+                // ================================================
+                let (assigned_keccak_values, assigned_sig_values): (
+                    Vec<[AssignedValue<F>; 4]>,
+                    Vec<AssignedSignatureVerify<F>>,
+                ) = signatures
+                    .iter()
+                    .chain(std::iter::repeat(&SignData::default()))
+                    .take(self.max_verif)
+                    .zip_eq(assigned_ecdsas.iter())
+                    .zip_eq(sign_data_decomposed.iter())
+                    .map(|((_, assigned_ecdsa), sign_data_decomp)| {
+                        self.assign_sig_verify(
+                            &mut ctx,
+                            &ecdsa_chip.range,
+                            sign_data_decomp,
+                            challenges,
+                            assigned_ecdsa,
+                        )
+                    })
+                    .collect::<Result<
+                        Vec<([AssignedValue<F>; 4], AssignedSignatureVerify<F>)>,
+                        Error,
+                    >>()?
+                    .into_iter()
+                    .unzip();
+
+                // ================================================
+                // step 4: deferred keccak checks
+                // ================================================
+                for (i, [is_address_zero, pk_rlc, pk_hash_lo, pk_hash_hi]) in
+                    assigned_keccak_values.iter().enumerate()
+                {
+                    let offset = i * 4;
+                    self.enable_keccak_lookup(
+                        config,
+                        &mut ctx,
+                        offset,
+                        is_address_zero,
+                        pk_rlc,
+                        &Word::new([*pk_hash_lo, *pk_hash_hi]),
+                    )?;
+                }
+
+                // IMPORTANT: this assigns all constants to the fixed columns
+                // IMPORTANT: this copies cells to the lookup advice column to perform range
+                // check lookups
+                // This is not optional.
+                let lookup_cells = ecdsa_chip.finalize(&mut ctx);
+                log::info!("total number of lookup cells: {}", lookup_cells);
+
+                ctx.print_stats(&["ECDSA context"]);
+                Ok(assigned_sig_values)
+            },
+        )?;
+
+        layouter.assign_region(
```

Comment:
There are some TODOs in the Scroll version, like `// TODO: is this correct?` before `assign_region` function. I guess it's a leftover?

Reply:
Yep, I believe so. So, I removed it.